### PR TITLE
[MERGE][REF] product, sale, *: clean pricelist API

### PR DESCRIPTION
**Pull request by Feyensv**
Refactor the whole sale price computation logic in `product` & `sale` to improve code quality, maintainability, ...

### Drop/reduce as much as possible context-based logic

* Drop magic `price` and `pricelist_id` fields on `product.product`/`product.template`
  * Keep one unique `_get_contextual_price()` to provide the old logic where it's still needed/difficult to remove atm.
* Do not rely on context fallbacks on the lower level of price computation (aka `price_compute` method)
  It is way too dangerous to have such magic fallbacks so deeply in the logic, inducing potentially unexpected/wrong results

### Clean `product.pricelist` API

* Clear parameters and fallbacks
  * Do not provide unused parameter(s) anymore (`partner`)
  * Stop providing products & quantities as a list of tuples, it was never used with different quantities in the existing code anyway
* Removed unused methods
* Split methods logic to ease and allow clean overrides and customizations
* Convert the old hardcoded SQL query to a clear domain, using ORM querying and formatting logic instead
* Improved documentation (docstrings, ...)

### Side cleanups in related/impacted modules

* Update to use the new clean pricelist API
* Code cleanup in sale prices computation (`sale_*`, `website_sale`, ...)
* Some onchanges -> computes conversion in `sale_*` scope
* Cache applied pricelist rule on SOL (avoid recomputing the pricelist rule in the price AND in the discount computation)

Task ID - 2688106
Community PR: https://github.com/odoo/odoo/pull/79605
Enterprise PR: https://github.com/odoo/enterprise/pull/22211

Co-authored-by: Victor Feyens <vfe@odoo.com>
Co-authored-by: Yannick Tivisse <yti@odoo.com>

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product/models/product_pricelist.py [hunk 1]
```diff
@@ -593,6 +559,59 @@ def write(self, values):
...
+        product_uom = product.uom_id
+        if uom != product_uom:
+            qty_in_product_uom = uom._compute_quantity(qty, product_uom, raise_if_failure=False)
+        else:
+            qty_in_product_uom = qty
+
+        for rule in self:
+            if rule.min_quantity and qty_in_product_uom < rule.min_quantity:
+                continue
+
```
Comment (Feyensv) [inline addons/product/models/product_pricelist.py:598]: After more thoughts on it, we should instead follow the idea of https://github.com/odoo/odoo/pull/81350 and keep the loop & qty_in_product_uom computation in `_compute_price_rule`

File: addons/sale_management/models/sale_order.py [hunk 2]
```diff
@@ -220,28 +231,41 @@ def _search_is_present(self, operator, value):
...
+
+    @api.depends('product_id', 'uom_id', 'quantity')
+    def _compute_price_unit(self):
+        for option in self:
+            if not option.product_id or not option.order_id.pricelist_id:
+                continue
+            # To compute the price_unit a so line is created in cache
+            values = option._get_values_to_add_to_order()
+            new_sol = self.env['sale.order.line'].new(values)
+            option.price_unit = new_sol._get_display_price()
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:257]:
```suggestion
            new_sol._compute_price_unit()
            option.price_unit = new_sol.price_unit
```

File: addons/sale_purchase/models/product_template.py [hunk 3]
```diff
@@ -18,9 +18,10 @@ class ProductTemplate(models.Model):
     ]
 
     @api.constrains('service_to_purchase', 'seller_ids')
-    def validate_service_to_purchase(self):
-        if self.service_to_purchase and not self.seller_ids:
-            raise ValidationError("Please define the vendor from whom you would like to purchase this service automatically.")
+    def _check_service_to_purchase(self):
+        for template in self:
+            if template.service_to_purchase and not template.seller_ids:
+                raise ValidationError("Please define the vendor from whom you would like to purchase this service automatically.")
```
Comment (Feyensv) [inline addons/sale_purchase/models/product_template.py:24]: Oh, didn't notice the translation was missing here :eyes:

File: addons/stock/models/product.py [hunk 4]
```diff
@@ -799,25 +802,28 @@ def _compute_product_tooltip(self):
     def _onchange_tracking(self):
         return self.mapped('product_variant_ids')._onchange_tracking()
 
+    @api.depends('type')
+    def _compute_tracking(self):
+        self.filtered(
+            lambda t: not t.tracking or t.type == 'consu' and t.tracking != 'none'
```
Comment (Feyensv) [inline addons/stock/models/product.py:808]: The `and t.tracking != 'none' is probably not necessary, unless writing the same value triggers a lot of other computes :thinking:

**Comment by tivisse** [comment]
@robodoo override=ci/security
Reason: False positive when adding another attribute on the field definition next to ``sanitize=False``

cc @mart-e

File: addons/product/models/product_pricelist.py [hunk 5]
```diff
@@ -74,239 +71,202 @@ def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_ui
...
+        rules = self._get_applicable_rules(date, product_tmpl_ids, product_ids, category_ids)
 
-            if not suitable_rule:
-                cur = product.currency_id
-                price = cur._convert(price, self.currency_id, self.env.company, date, round=False)
+        if 'currency' in products.env.context:
+            # Remove currency from product context to avoid any
+            # side-effect currency conversion in price_compute calls
+            # since we manage currency conversion without the context fallbacks
+            product = products.with_context(currency=False)
```
Comment (Feyensv) [inline addons/product/models/product_pricelist.py:175]: might be better to pop the key from the ctxt instead :thinking:

File: addons/sale_management/models/sale_order.py [hunk 6]
```diff
@@ -88,8 +88,9 @@ def _compute_option_data_for_template_change(self, option):
     def update_prices(self):
         self.ensure_one()
         res = super().update_prices()
+        # TODO env.add_to_compute after onchange conversion to computes
         for line in self.sale_order_option_ids:
-            line.price_unit = self.pricelist_id.get_product_price(line.product_id, line.quantity, self.partner_id, uom_id=line.uom_id.id)
+            line.price_unit = self.pricelist_id._get_product_price(line.product_id, line.quantity, uom=line.uom_id)
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:93]: self.sale_order_option_ids._compute_price_unit()
self.sale_order_option_ids._compute_discount()

Comment (Feyensv) [reply addons/sale_management/models/sale_order.py:93]: Also remove the
```python
        if not is_html_empty(template.note):
            self.note = template.note
```

below in the onchange, it is managed by the compute

**Comment by Feyensv** [comment]
@robodoo rebase-merge

**Comment by tivisse** [comment]
@robodoo r+ merge

**Pull request merged**